core/src/auth_service/utility.bal

```diff
@@ -0,0 +1,105 @@
+import ballerina/config as conf;
+import ballerina/crypto;
+import ballerina/jwt;
+import ballerina/time;
+
+# The `generateJWT` will generate the JWT token for the given user credentials.
+# 
+# + credentials - User credentials.
+# + return - This function will return either JWT or an appropriate error
+function generateJWT(json credentials) returns string|error {
+
+    [string, string] [email, password] = check extractCredentials(credentials);
+    json|error authenticateUserResult = authenticateUser(email, password);
+    if (authenticateUserResult is json) {
+
+        // Get id and userType.
+        string id = check trap <string>authenticateUserResult.id;
+        string scope = check trap <string>authenticateUserResult.userType;
+
+        // keyStore setup.
+        crypto:KeyStore keyStore = {
+            path: "resources/ballerinaKeystore.p12",
+            password: conf:getAsString("KEY_STORE_PASSWORD")
+        };
+
+        jwt:JwtKeyStoreConfig keyStoreConfig = {
+            keyStore: keyStore,
+            keyAlias: "ballerina",
+            keyPassword: "ballerina"
+        };
+
+        // Set JWT header.
+        jwt:JwtHeader header = {};
+        header.alg = jwt:RS256;
+        header.typ = "JWT";
+
+        // Set JWT payload.
+        jwt:JwtPayload payload = {};
+        payload.sub = "Auth token";
+        payload.iss = "environment platform";
+        payload.jti = "100078234ba23";
+        payload.customClaims = {name: id, scope: scope};
+        payload.exp = time:currentTime().time / 1000 + conf:getAsInt("JWT_EXPIRE", 3600);
+        payload.iat = time:currentTime().time / 1000;
+        string jwt = check jwt:issueJwt(header, payload, keyStoreConfig);
+        return jwt;
+    } else {
+        return authenticateUserResult;
+    }
+}
+
+# The `getHashedPassword` will convert the given password to SHA265 encoded string.
+# 
+# + password - Password to be hashed.
+# + return - This function will return the hashed value of the given password using SHA256 with base 16.
+function getHashedPassword(string password) returns string {
+
+    byte[] passwordByteArray = password.toBytes();
+    byte[] hashedPassword = crypto:hashSha256(passwordByteArray);
+    return hashedPassword.toBase16();
+}
+
+# The `constructUserInformation` will construct user information into a json.
+# 
+# + id - Id of the user.
+# + userType - UserType of the user.
+# + return - This function will return the id and the usertype as a json.
+function constructUserInformation(string id, string userType) returns json {
+    return {id: id, userType: userType};
+}
+
+# The `authenticateUser` function will get the Id and the userType of the user if valid email and password given.
+# 
+# + email - Email of the user.
+# + password - Password of the user.
+# + return - This function will either return the authenticated user Id and the user type or an appropriate error.
+function authenticateUser(string email, string password) returns json|error {
```

Comment:
I'm a little concerned how error handling will happen here. If a user is not found in getUser(), getAdmin() or getMinistryEmployee(), an error will be thrown right? @yashodgayashan

Reply:
Yeah @sanjulamadurapperuma it will either return the error or the json of the authenticated user.

Reply:
yes.

Reply:
In that case, would the authentication flow break since there is no check for the type of user in this method? Like for example, there can be a generic user (not admin or ministry employee). When that user tries to sign in, an error will be thrown all the time because an error is thrown when the getAdmin() method is invoked (assuming method invocation stops from there). Correct me if I am wrong.

If this is the case, I would suggest passing in the type of user into the authenticateUser method so that only the relevant method will be invoked (getUser, getAdmin or getMinistryEmployee).

Reply:
` string hashedPassword = getHashedPassword(password);
    json|error user = getUser(email, hashedPassword);
    if (user is json || user.reason() == "Incorrect password") {
        return <@taint>user;
    }
    json|error admin = getAdmin(email, hashedPassword);
    if (admin is json || admin.reason() == "Incorrect password") {
        return <@taint>admin;
    }
    json|error ministry = getMinistryEmployee(email, hashedPassword);
    if (ministry is json || ministry.reason() == "Incorrect password") {
        return <@taint>ministry;
    }
    return error("No user found", message = "Couldn't find the user with the given credentials.");`

No it will check if user is there or password incorrect if there is no user in generic user, admin or ministry then, it will throw the error.

Reply:
```
    string hashedPassword = getHashedPassword(password);
    json|error user = getUser(email, hashedPassword);
    if (user is json || user.reason() == "Incorrect password") {
        return <@taint>user;
    }
    json|error admin = getAdmin(email, hashedPassword);
    if (admin is json || admin.reason() == "Incorrect password") {
        return <@taint>admin;
    }
    json|error ministry = getMinistryEmployee(email, hashedPassword);
    if (ministry is json || ministry.reason() == "Incorrect password") {
        return <@taint>ministry;
    }
    return error("No user found", message = "Couldn't find the user with the given credentials.");
```

No it will check if user is there or password incorrect if there is no user in generic user, admin or ministry then, it will throw the error.

Reply:
That's my point, if a user record exist in getUser, but that same record does not exist in Admin or ministry user, then the errors thrown from those methods will disrupt the flow in authenticateUser right?

Reply:
Because according to the current flow in the authenticateUser method, all three methods get invoked regardless of whether the user exists in those tables

Reply:
No. If user found in the getUser method it will return if not it will jump to the get admin. Likewise it go to getMinistry if user is not found in three tables only send the error that user is not found.

Reply:
Okay, I had a misunderstanding, I think this error flow is handled already

Reply:
Thanks for the clarification

Reply:
Yes, I think. If there are any doubts please ask. The flow is,
* first using the get user method it will take the error or the user json.
* If json or the error is incorrect password it will return or else it will check from the get admin.
* likewise it check three tables and if user is not found it will pass the error saying user is not found.

Hope this will help.